### ISSUE TITLE
RFD: waitOn and data

### ISSUE THREAD

**Issue by jamgold** [EventedMind/iron-router#1193]
This is a Request For Documentation. I just burnt two hours and finally figured out that when you have a waitOn or subcriptions in your route, and also a data function, you have to manually make sure the route is ready

```
Router.route('something',{
 waitOne: function() {
  return Meteor.subscribe('somehugecollection');
 },
 data: function() {
  if(this.ready())
  {
   return SomeHugeCollection.find();
  }
 }
})
```

otherwise the template will be rendered incomplete, nor will the loading plugin/loadingTemplate work correctly.

I could not find this mentioned in the documentation.

**Comment by vladblindu** [comment]
Yep, same issue here. I needed a count() of the  cursor available through data function and it shows 0, although later, being reactive, the table renders correctly.
```
Router.onBeforeAction('loading');

this.route("admin_db",{
		path:"/admin/:area",
		layoutTemplate: 'adminLanding',
		template: "dataTable",
		waitOn: function(){
			Meteor.subscribe("admin_" + capitalize(this.params.area));
		},
		data: function(){
			var collectionData = AppCollections[capitalize(this.params.area)].find();
			return {
				collectionData: collectionData,
				title:capitalize(this.params.area)
			}
		}
	});
```
I need to know form start how many items are in the subscriber cursor to do some styling, which is impossible as long as in the template.dataTable.rendered i always get count() = 0.
(by the way it's not a huge collection, it has 4 documents?!?)

**Comment by dcsan** [comment]
yeh i'm in the middle of this process. the router docs seem to be worse than pre 1.0; esp since the API changed so most of the google hits are incorrect.
there is no good example of how to use waitOn, data and if the `action` is needed.

@jamgold you also have a typo: waitOn(e)

**Comment by jamgold** [comment]
@dcsan thanks, corrected typo

**Comment by kk4sh** [comment]
Im quite new in meteor  but  also quite stuck with the good way to use waitOn/subscriptions data....
Even with a light example, the render/subscriptions seems long and transistion between data is clearly not smooth enough in such a context...
[http://meteorpad.com/pad/bSv6KofP8NrDhA2r4/Leaderboard](http://meteorpad.com/pad/bSv6KofP8NrDhA2r4/Leaderboard)
Of course, the final render decreases when the page template is more complex.
Is there a way to be faster/cleaner or is it a "normal" behavior in such a use case? 
Is there somewhere to find the good way to do it ?

**Comment by hamoid** [comment]
I also just wasted 2 or 3 hours trying to debug this. It's a rather large project and things were working fine. Then I added one route and it would just not render. The subscribed collections were printed as empty inside data(). But in the console I could query and see the data was there, very puzzling.

Wrapping the content of ```data()``` in a ```if(this.ready())``` fixed the issue. Thanks for that tip! I wonder why I need this now: other routes don't use that trick and work fine.

**Comment by proehlen** [comment]
Same issue and had to apply the same workaround as hamoid.  Thought it might have somethign to do with multiple subscriptions in waitOn but maybe that's not the case.  Requiring an `if(this.ready())` seems to be redundant according to my understanding of what waitOn is supposed to do.

**Comment by Tzaphkiel** [comment]
I believe it has to do with multiple subscriptions indeed. I put a counter in the router that resets on the
```
return [Meteor.subscribe(...), Meteor.subscribe(...)]
```
and however many items I put in there, the data method gets called multiple times, i guess when each of the subscription is met.

The ```if(this.ready())``` workaround only triggers the counter update once when all ```n``` subscriptions are met...

This should be fixed in the waitOn method and I believe it should be considered a bug...no ?


Example route:
```
Router.route('/:fromCount?', {
  name: 'eventsList',
  waitOn: function() {
    var l = Session.get("EventsFromCount") || 0;
    var filterOptions = Session.get("eventsFilter") || {};
    this.c = 0;
    return [
      Meteor.subscribe('events', l, getSortOptions(), filterOptions),
      Meteor.subscribe('eventStats'),
      Meteor.subscribe('eventsCount', filterOptions)
    ];
  },
  data: function() {
    if(this.ready()){
      this.c += 1;
      Meteor.call("debug", this.c + ") NumEvents gotten: " + Events.find().count());
      return Events.find({}, {sort: getSortOptions()});
    }
  }
})
```
